### PR TITLE
Refactor Transitive Order Reducing Loops

### PR DESCRIPTION
**Pull request by nlordell**
This PR introduces a new graph primitive operation to conditionally fill the optimal path between a token pair and return the flow along the path. This allows some refactoring and simplifies the implementation of some higher level methods.

### Test Plan

All unit tests continue to pass without modifications :tada:

### CODE REVIEW

File: pricegraph/src/orderbook.rs [hunk 1]
```diff
@@ -285,112 +262,146 @@ impl Orderbook {
...
+    /// not check a condition on the discovered path's flow before filling.
+    fn fill_optimal_transitive_order(
+        &mut self,
+        pair: TokenPair,
+        reduce_negative_cycles: bool,
+    ) -> Option<Flow> {
+        self.fill_optimal_transitive_order_if(pair, reduce_negative_cycles, |_| true)
+    }
+
+    /// Fills the optimal transitive order for the specified token pair by
```
Comment (e00E) [inline pricegraph/src/orderbook.rs:354]: Define "optimal" in comment

File: pricegraph/src/orderbook.rs [hunk 2]
```diff
@@ -285,112 +262,146 @@ impl Orderbook {
...
+    /// This is calculated by repeatedly finding the optimal counter transitive
+    /// orders whose limit exchange rates overlap with the specified limit
+    /// exchange rate.
     ///
-    /// Note that the limit price is expressed as an exchange limit price, i.e.
-    /// with implicitely included fees. Additionally, the invariant
-    /// `buy_amount / sell_amount <= limit_price` holds, but in general the
-    /// ratio of the two amounts does not equal the limit price.
-    pub fn fill_order_at_price(&mut self, pair: TokenPair, limit_price: f64) -> (f64, f64) {
+    /// Note that the limit exchange rate implicitely included fees.
```
Comment (e00E) [inline pricegraph/src/orderbook.rs:307]:
```suggestion
    /// Note that the limit exchange rate implicitly included fees.
```

**Review by e00E — APPROVED** [review]
Good refactoring

**Review by fleupold — APPROVED** [review]
<a href="https://gitme.me/image?url=https%3A%2F%2Fmedia2.giphy.com%2Fmedia%2F3o6vY0ucbVFbNkym1q%2Fgiphy.gif&token=loving_it" data-gitmeme-token="loving_it"></a>

**Pull request merged**